src/openrct2-ui/windows/PatrolArea.cpp

```diff
@@ -0,0 +1,304 @@
+            return;
+
+        const auto res = String::Parse<int32_t>(text);
+        if (res.has_value())
+        {
+            int32_t size;
+            size = res.value();
+            size = std::max(MINIMUM_TOOL_SIZE, size);
+            size = std::min(MAXIMUM_TOOL_SIZE, size);
+            gLandToolSize = size;
+            Invalidate();
+        }
+    }
+
+    void OnUpdate() override
+    {
+        // Close window if another tool is open or staff window gets closed
+        if (!PatrolAreaToolIsActive() || !IsStaffWindowOpen())
+        {
+            Close();
+        }
+    }
+
+    void OnPrepareDraw() override
+    {
+        SetWidgetPressed(WIDX_PREVIEW, true);
+        PatrolAreaWidgets[WIDX_PREVIEW].image = LandTool::SizeToSpriteIndex(gLandToolSize);
+    }
+
+    void OnDraw(rct_drawpixelinfo& dpi) override
+    {
+        DrawWidgets(dpi);
+
+        // Draw number for tool sizes bigger than 7
+        if (gLandToolSize > MAX_TOOL_SIZE_WITH_SPRITE)
+        {
+            auto screenCoords = ScreenCoordsXY{ windowPos.x + PatrolAreaWidgets[WIDX_PREVIEW].midX(),
+                                                windowPos.y + PatrolAreaWidgets[WIDX_PREVIEW].midY() };
+            auto ft = Formatter();
+            ft.Add<uint16_t>(gLandToolSize);
+            DrawTextBasic(&dpi, screenCoords - ScreenCoordsXY{ 0, 2 }, STR_LAND_TOOL_SIZE_VALUE, ft, { TextAlignment::CENTRE });
+        }
+    }
+
+    void OnToolUpdate(rct_widgetindex widgetIndex, const ScreenCoordsXY& screenCoords) override
+    {
+        auto mapTile = GetBestCoordsFromPos(screenCoords);
+        if (!mapTile)
+            return;
+
+        auto stateChanged = false;
+        if (!(gMapSelectFlags & MAP_SELECT_FLAG_ENABLE))
+            stateChanged = true;
+
+        if (gMapSelectType != MAP_SELECT_TYPE_FULL)
+            stateChanged = true;
+
+        auto toolSize = std::max<uint16_t>(1, gLandToolSize);
+        auto toolLength = (toolSize - 1) * 32;
+
+        // Move to tool bottom left
+        mapTile->x -= (toolSize - 1) * 16;
+        mapTile->y -= (toolSize - 1) * 16;
+        mapTile = mapTile->ToTileStart();
+        auto posA = *mapTile;
+        mapTile->x += toolLength;
+        mapTile->y += toolLength;
+        auto posB = *mapTile;
+        if (gMapSelectPositionA != posA || gMapSelectPositionB != posB)
+            stateChanged = true;
+
+        if (stateChanged)
+        {
+            // Invalidate previous area
+            map_invalidate_selection_rect();
+
+            // Update and invalidate new area
+            gMapSelectFlags |= MAP_SELECT_FLAG_ENABLE;
+            gMapSelectType = MAP_SELECT_TYPE_FULL;
+            gMapSelectPositionA = posA;
+            gMapSelectPositionB = posB;
+            map_invalidate_selection_rect();
+        }
+    }
+
+    void OnToolAbort(rct_widgetindex widgetIndex) override
+    {
+        hide_gridlines();
+        ClearPatrolAreaToRender();
+        gfx_invalidate_screen();
```

Comment:
Should this window auto close on tool abort?

Reply:
It does